components/proxy/src/main/kotlin/com/hotels/styx/routing/db/StyxObjectStore.kt

```diff
@@ -15,28 +15,100 @@
  */
 package com.hotels.styx.routing.db;
 
-import com.hotels.styx.api.configuration.ObjectStore
+import com.hotels.styx.api.configuration.ObjectStoreSnapshot
+import org.pcollections.HashTreePMap
+import org.pcollections.PMap
+import org.reactivestreams.Publisher
+import reactor.core.publisher.Flux
 import java.util.Optional
-import java.util.concurrent.ConcurrentHashMap
+import java.util.concurrent.CopyOnWriteArrayList
+import java.util.concurrent.Executors
+import java.util.concurrent.atomic.AtomicReference
+import java.util.function.Consumer
 
 /**
  * Styx Route Database.
  */
-class StyxObjectStore<T> : ObjectStore<T> {
-    private val objects = ConcurrentHashMap<String, Record<T>>();
+class StyxObjectStore<T> : ObjectStoreSnapshot<T> {
+    private val objects: AtomicReference<PMap<String, Record<T>>> = AtomicReference(HashTreePMap.empty())
+    private val watchers = CopyOnWriteArrayList<ChangeWatcher<T>>()
+    private val executor = Executors.newSingleThreadExecutor()
```

Comment:
This Executor will never be stopped, this seems relevant for tests where we could instantiate the class several times. If we think that production code 'should be ok since there's a single StyxObjectStore that will exist for the whole lifetime of the application', it makes sense to convert this class into an object so the singleton nature is depicted.

Reply:
How about, if for now, I'd just declare `executor` in the companion object? Like so:
    
    companion object {
        val executor = Executors.newSingleThreadExecutor()
    }

This addresses thread leak in tests, and will work well in production code. Also this executor is likely to have a very low work load, it would work sufficiently well in production code even with multiple ObjectStore implementations/instances.

Reply:
It sounds safe enough (unless the test framework uses several classloaders to load the class! of course).